common/core/desktop/src/kmx/kmx_xstring.cpp

```diff
@@ -116,37 +116,43 @@ km_kbp_cp *km::kbp::kmx::u16tok(km_kbp_cp *p, km_kbp_cp ch, km_kbp_cp **ctx) {
 *
 * xstrlen calculates the length of a string, ignoring some special chars.
 */
+PKMX_WCHAR km::kbp::kmx::incxstr(PKMX_WCHAR p) {
 
-PKMX_WCHAR km::kbp::kmx::incxstr(PKMX_WCHAR p)
-{
-  if(*p == 0) return p;
-  if(*p != UC_SENTINEL)
-  {
-    if(*p >= 0xD800 && *p <= 0xDBFF && *(p+1) >= 0xDC00 && *(p+1) <= 0xDFFF) return p+2;
-    return p+1;
+  if (*p == 0)
+    return p;
+  if (*p != UC_SENTINEL) {
+    if (*p >= 0xD800 && *p <= 0xDBFF && *(p + 1) >= 0xDC00 && *(p + 1) <= 0xDFFF)
+      return p + 2;
+    return p + 1;
+  }
+  // UC_SENTINEL(FFFF) with UC_SENTINEL_EXTENDEDEND(0x10) == variable length
+  if (*(p + 1) == CODE_EXTENDED) {
+    p += 2;
+    while (*p && *p != UC_SENTINEL_EXTENDEDEND)
+      p++;
+    if (*p == 0)
+      return p;
+    return p + 1;
   }
 
-  p+=2;
-  switch(*(p-1))
-  {
-    case CODE_ANY:      return p+1;
-    case CODE_NOTANY:   return p+1;
-    case CODE_INDEX:    return p+2;
-    case CODE_USE:      return p+1;
-    case CODE_DEADKEY:    return p+1;
-    case CODE_EXTENDED:   p += 2; while(*p && *p != UC_SENTINEL_EXTENDEDEND) p++; return p+1;
-    case CODE_CLEARCONTEXT: return p+1;
-    case CODE_CALL:     return p+1;
-    case CODE_CONTEXTEX:  return p+1;
-    case CODE_IFOPT:    return p+3;
-    case CODE_IFSYSTEMSTORE: return p+3;
-    case CODE_SETOPT:   return p+2;
-    case CODE_SETSYSTEMSTORE: return p+2;
-    case CODE_RESETOPT: return p+1;
-    case CODE_SAVEOPT:  return p+1;
-    default:        return p;
+  // CODE_PTR defined in kmx_processevent.h
+  for (int i = 0; i < size(CODE_PTR[0]); i++) {
+    if (*(p + 1) == CODE_PTR[0][i]) {
+      deltaptr = CODE_PTR[1][i];
+      break;
+    }
   }
```

Comment:
I think we should be able to do this without needing a loop. If we define a `CODE__SIZE` array in kmx_file.h, e.g.:

```
/// number of parameters for each CODE_### control code
extern const int CODE__SIZE[];
```

and then in kmx_file.cpp (copying from the #defines for CODE_### in kmx_file.h):

```
#include <assert.h>
...

/// number of parameters for each CODE_### control code
const int CODE__SIZE[] = {
  -1, // undefined       0x00
  1,  // CODE_ANY        0x01
  2,  // CODE_INDEX      0x02
  0,  // CODE_CONTEXT    0x03
...
  // ... use 0 for undefined CODE_ values, e.g. 0x09
};

// Ensure that all CODE_### sizes are defined
static_assert(sizeof(CODE__SIZE)/sizeof(CODE_SIZE_[0]) == CODE_LASTCODE + 1);
```

then we can do, in `incxstr`: 

```
if(*p > CODE_LASTCODE || CODE__SIZE[*p] == -1) {
  // Invalid CODE_ value
  return *p;
}

int deltaptr = CODE__SIZE[*p];
p++;
//...
```

Reply:
this is included in todays commit